### PR TITLE
esp_rom_flash: handle read/write alignment

### PR DESCRIPTION
**Pull request by sylvioalves**
Check alignment before proceeding with operation.
If src, dest or length are not 4-byte align, use a temp buffer to handle it.

### CODE REVIEW

File: zephyr/port/bootloader/bootloader_flash.c [hunk 1]
```diff
@@ -30,6 +30,13 @@
 
 #include "esp_rom_spiflash.h"
 
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+
+#define DT_DRV_COMPAT espressif_esp32_flash_controller
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+#define FLASH_BLK_SZ DT_PROP(SOC_NV_FLASH_NODE, write_block_size)
```
Comment (sylvioalves) [inline zephyr/port/bootloader/bootloader_flash.c:38]: This is usually small buffer.. up to 32 bytes. We could increase it to 256 byte-long as well.

**Comment by sylvioalves** [comment]
OTA tests ok!

**Review by wmrsouza — APPROVED** [review]

File: zephyr/port/bootloader/bootloader_flash.c [hunk 2]
```diff
@@ -333,53 +340,93 @@ static esp_err_t bootloader_flash_read_allow_decrypt(size_t src_addr, void *dest
...
-    if (allow_decrypt) {
-        return bootloader_flash_read_allow_decrypt(src_addr, dest, size);
-    } else {
-        return bootloader_flash_read_no_decrypt(src_addr, dest, size);
+    /* handle misaligned source address */
+    size_t src_offset = src_addr & 3;
+    size_t start_addr = src_addr - src_offset;
+
+    while (remaining > 0) {
+        copy_size = (remaining > FLASH_BLK_SZ) ? FLASH_BLK_SZ : remaining;
```
Comment (martinjaeger) [inline zephyr/port/bootloader/bootloader_flash.c:353]: This should use `sizeof`, just in case the `temp_buf`'s size is changed in the future.

```suggestion
        copy_size = (remaining > sizeof(temp_buf)) ? sizeof(temp_buf) : remaining;
```

File: zephyr/port/bootloader/bootloader_flash.c [hunk 3]
```diff
@@ -333,53 +340,93 @@ static esp_err_t bootloader_flash_read_allow_decrypt(size_t src_addr, void *dest
...
+        aligned_size = (copy_size + src_offset + 3) & ~3;
+
+        /* read from flash into the temp buffer */
+        esp_err_t err;
+        if (allow_decrypt) {
+            err = bootloader_flash_read_allow_decrypt(start_addr, temp_buf, aligned_size);
```
Comment (martinjaeger) [inline zephyr/port/bootloader/bootloader_flash.c:359]: `aligned_size` could be more than the buffer size, leading to buffer overflow.

Example (in first run of the loop):
```
src_addr = 3;
src_offset = 3;
start_addr = 0;
copy_size = 32;
aligned_size = (32 + 3 + 3) & ~3 = 36;
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]
I think we need some tests to check these alignment calculations. It's quite difficult to be absolutely sure that all edge cases are covered just by looking at the code.

If I didn't get it wrong, the current implementation might again lead to a buffer overflow.

File: zephyr/port/bootloader/bootloader_flash.c (hunk 3 above)
Comment (sylvioalves) [reply zephyr/port/bootloader/bootloader_flash.c:359]: Perhaps we can add the max offset which would be 3.. then rouding to 4:
`uint8_t temp_buf[FLASH_BLK_SZ + 4];`

**Comment by martinjaeger** [comment]
I'm wondering why we actually need an implementation that allows unaligned writes. I've looked at the STM32 driver in Zephyr upstream and it would just return an error code if alignment is not correct. Can't we do the same here and avoid the difficult automatic alignment implementation?
